Basic integration with ClickHouse

vavrusa opened the pull request:
This PR adds integration with ClickHouse SQLAlchemy engine spec.

* [x] Adding database and tables works
* [x] Adding views works
* [x] Table, timeseries, group by, group by + timeseries works
* [x] Using secondary dttm columns (ClickHouse has often `date` + `datetime`, so timeseries needs to reference both)
* [x] Superset makes quoted (e.g. `AVG("column")`) metrics by default when it shouldn't, don't know why
* [x] Filter on INTEGER fields creates `column IN ("1", "2", ...)` SQL where values are quoted, this is wrong
* [x] Superset launches a query as soon as I create slice, this is bad as it potentially full-scans the whole database

mistercrunch commented:
I think you forked off a broken `master`, mind rebasing?

vavrusa commented:
This looks okay now, not sure if you want to cover ClickHouse with system tests, as that would require installing it as a dependency for tests.

goranc commented:
Any progress with merge to master ?

dmzaytsev commented:
@vavrusa @mistercrunch integration with ClickHouse is a really good feature
could you guys continue work on merge to master
thank you

vavrusa commented:
I've rebased again, not sure what else can I do.

dmzaytsev commented:
@vavrusa thank you

@mistercrunch please take a look
will you accept this PR?

ascott commented on superset/assets/visualizations/filter_box.jsx:
```diff
@@ -72,26 +72,38 @@ class FilterBox extends React.Component {
+      const val = this.state.selectedValues[filter];
+      if (Array.isArray(val)) {
+        val.forEach(function (v) {
+          const exists = data.find(function (e) {
```
could use fat arrow syntax here for consistency `(e) => {}`

ascott replied:
could also simplify this line to: `const exists = data.find((e) => e.id === v);`

ascott commented on superset/assets/visualizations/filter_box.jsx:
```diff
@@ -72,26 +72,38 @@ class FilterBox extends React.Component {
       });
+      const val = this.state.selectedValues[filter];
+      if (Array.isArray(val)) {
+        val.forEach(function (v) {
```
could use fat arrow here too.

vavrusa replied:
"fat arrow" means the `v => {}` thingy?

ascott replied:
yep, that's right `(v) => {}`

ascott commented on superset/assets/visualizations/filter_box.jsx:
```diff
@@ -72,26 +72,38 @@ class FilterBox extends React.Component {
+          const exists = data.find(function (e) {
+            return e.id === v;
+          });
+          if (exists === undefined) {
```
any reason not to use `if (!exists) { }`?

ascott commented on superset/assets/visualizations/filter_box.jsx:
```diff
@@ -72,26 +72,38 @@ class FilterBox extends React.Component {
+      }
+      const options = data.map((opt) => {
+        const perc = Math.round((opt.metric / maxes[opt.filter]) * 100);
+        const backgroundImage = (
```
could you please add a screenshot of this ui change?

ascott replied:
nevermind, i see you just moved this from another line.

mistercrunch commented on superset/models.py:
```diff
@@ -1295,7 +1300,7 @@ def query(  # sqla
             timeseries_limit_metric_expr = \
                 timeseries_limit_metric.sqla_col
-        if metrics:
+        if metrics and len(metrics_exprs) > 0:
```
I think this could be simplified to just `if metrics_exprs:`

vavrusa replied:
Indeed

ascott commented:
hi @vavrusa, :wave:, thanks for this PR! sorry for the wait on getting it merged. i know @mistercrunch is busy preparing for a talk he's giving tomorrow but hopefully he can chime in on these changes tomorrow.

mistercrunch commented on superset/models.py:
```diff
@@ -1346,13 +1352,27 @@ def visit_column(element, compiler, **kw):
-            time_filter = dttm_col.get_time_filter(from_dttm, to_dttm)
+            # Use all dttm columns to support index with secondary dttm columns
+            dttm_col_all = False
+            if hasattr(self.database.db_engine_spec, 'time_secondary_columns'):
```
let's set `time_secondary_columns = False` in the base class

vavrusa replied:
Done

mistercrunch replied:
Now you don't need the `hasattr` anymore and perhaps not `dttm_col_all`, just use time_secondary_columns

mistercrunch commented:
Can you describe how the time indexes/grouping work in Clickhouse so that we understand this PR?

mistercrunch commented:
"Superset makes quoted (e.g. AVG("column")) metrics by default when it shouldn't, don't know why"

This quoting should be specified in your SQLAlchemy dialect

vavrusa commented:
> Can you describe how the time indexes/grouping work in Clickhouse?

The most common table type [MergeTree](https://clickhouse.yandex/reference_en.html#MergeTree) requires segmentation by `Date` (number of days since epoch), but allows you to add other columns, like `DateTime` for seconds. These columns are however independent, so constraining query like `datetime  > now() - 3600` doesn't automatically infer that column `date` is either today or yesterday, so you have to place constraints on both columns like `datetime > now() - 3600 AND date = today()`.
Superset by default selects only main dttm column and places constraint on it, so if `datetime` is the main column, it will still inspect all parts segmented by `date` searching for `datetime`, instead of skipping parts where `date BETWEEN toDate(constraint) AND toDate(constraint)`. Does that make sense?

> This quoting should be specified in your SQLAlchemy dialect

Indeed, that's fixed in 617ea1153e34a1d6cc3b609dfc162748127c02f2

vavrusa commented on superset/models.py:
```diff
@@ -1527,8 +1527,8 @@ def fetch_metadata(self):
-            quoted = "{}".format(
-                column(dbcol.column_name).compile(dialect=db.engine.dialect))
+            engine = self.database.get_sqla_engine()
+            quoted = col.compile(dialect=engine.dialect)
```
^ The previous code created a column using a column name and compiled it, for some reason dialect override doesn't kick in. The new code takes existing column and compiles it using the dialect, that works.

vavrusa commented on superset/models.py:
```diff
@@ -1410,7 +1421,8 @@ def visit_column(element, compiler, **kw):
 
-        if is_timeseries and timeseries_limit and groupby:
+        # Some SQL dialects use timeslot grouping inline with other dimensions
+        if is_timeseries and timeseries_limit and groupby and not time_groupby_inline:
```
^ So the previous code rewritten `GROUP BY` as inner query, and then applied ordering and column selection in the outer query. I have zero idea why Vertica requires that for time groupings. It's also slower if the DB is unable to unroll the inner query, so it has to execute the inner query into temporary table and join it with the full table before ordering. CH treats time grouping as any other dimension, so you can use it in `ORDER BY` without problems. The resulting SQL looks just like any other grouping: `SELECT datetime, url, count() as cc FROM table GROUP BY datetime, url ORDER BY datetime, cc`

mistercrunch replied:
I don't think I understand this statement and I know the Vertica optimizer intimately.

vavrusa replied:
I guess this is because of this comment:

```
            # some sql dialects require for order by expressions
            # to also be in the select clause -- others, e.g. vertica,
            # require a unique inner alias
```

Otherwise I don't know why "some backends" need a subquery, the only thing I did is to remove the need for subquery for CH backend with a class variable.

markrey commented:
Any updates on the PR? This is quite amazing integration!

mistercrunch commented on superset/assets/visualizations/filter_box.jsx:
```diff
@@ -4,7 +4,7 @@ import d3 from 'd3';
 import ReactDOM from 'react-dom';
 
-import Select from 'react-select';
+import { Creatable } from 'react-select';
```
Wait do the changes on filter_box belong in this PR? Bad merge conflict resolve?

mistercrunch commented on superset/assets/webpack.config.js:
```diff
@@ -50,7 +50,7 @@ const config = {
         test: /\.jsx?$/,
         exclude: APP_DIR + '/node_modules',
-        loader: 'babel',
+        loader: 'babel-loader',
```
does this belong in this PR?

mistercrunch commented on superset/forms.py:
```diff
@@ -493,7 +493,7 @@ def __init__(self, viz):
             'since': (FreeFormSelectField, {
                 "label": _("Since"),
-                "default": "7 days ago",
+                "default": config.get('SUPERSET_DEFAULT_SINCE', '7 days ago'),
```
forms.py is no longer...

mistercrunch commented on superset/models.py:
```diff
@@ -1278,6 +1278,11 @@ def query(  # sqla
             granularity = self.main_dttm_col
 
+        # Database spec supports join-free timeslot grouping
+        time_groupby_inline = False
```
this whole section is not necessary, as we can use `self.database.db_engine_spec.time_groupby_inline` in place of the local var

vavrusa replied:
Ok

mistercrunch commented on superset/models.py:
```diff
@@ -1345,13 +1351,27 @@ def visit_column(element, compiler, **kw):
 
-            if is_timeseries:
+            # If primarily grouping by time without join, place time groupings first
+            if is_timeseries and time_groupby_inline:
```
```
if timeseries:
    timestamp = dttm_col.get_timestamp_expression(time_grain)
    if self.database.db_engine_spec.time_groupby_inline:
```
```

mistercrunch commented on superset/models.py:
```diff
@@ -1345,13 +1351,27 @@ def visit_column(element, compiler, **kw):
+            # If primarily grouping by time without join, place time groupings first
+            if is_timeseries and time_groupby_inline:
+                timestamp = dttm_col.get_timestamp_expression(time_grain)
+                select_exprs = [timestamp] + select_exprs
```
why is the ordering of the columns important?

mistercrunch commented on superset/models.py:
```diff
@@ -1345,13 +1351,27 @@ def visit_column(element, compiler, **kw):
+                dttm_col_all = self.database.db_engine_spec.time_secondary_columns
+            if dttm_col_all:
+                for c in self.dttm_cols:
+                    if c in cols:
```
why would c not be in cols, can that happen?

mistercrunch commented:
Please rebase! Sorry it's been quite a ride. We're pretty careful with things that touch the `query` method.

dmzaytsev commented:
@vavrusa could you please continue?
it's amazing feature

vavrusa commented:
@dmzaytsev Oops, sorry, I'll try to fixup things over the weekend. Feel free to help with the PR if you can. To be honest, I stopped using Superset because it's too buggy to use in production, so I don't have too much time to spend on it.

dmzaytsev commented:
@vavrusa it would be great, thanks. 
Unfortunately, I can't help you with this PR, as I am not a python(superset) expert
By the way, what are you using instead of superset?

mistercrunch commented:
@vavrusa for what's it's worth we have 700+ weekly active (and it's growing quickly) at Airbnb so we're going to focus more on quality moving forward. It's an act of balance between adding functionality and stabilizing it, we're starting to focus more on the quality/UX for sure though.

We're still very ambitious and bullish and want to keep moving fast, so for folks looking for perfection it may not be the right choice at the moment.

vavrusa commented:
@mistercrunch I didn't mean it in a bad way, it's just doesn't work for my use case. I'm sure it is fine for smaller datasets or teams. It's a 0.x version after all and I'm sure it'll be much better over time.
@dmzaytsev I'm using Grafana with panel extensions (histogram, heatmap, datatable, pie, sunburst) currently with https://github.com/vavrusa/grafana-sqldb-datasource datasource to enable ClickHouse.

vavrusa committed:
db_engine_specs: added engine spec for clickhouse

vavrusa committed:
models: group by time without join on subquery

if this is enabled, it will remove the subquery when grouping by time,
and group by time will be rewritten as group by timeslot, ...

this particularly fixes clickhouse backend

vavrusa committed:
docs: added clickhouse

vavrusa committed:
models: allow using secondary dttm columns as constraints

vavrusa committed:
superset/forms: configurable default slice since time

vavrusa committed:
superset/models: respect dialect quoting

vavrusa commented:
I rebased everything and removed some stuff I wasn't sure where to put. The basic functionality should be there.

RichRadics committed:
dttm handling, codeclimate tweaks

RichRadics commented:
Rebased to current master, and reformatted to satisfy codeclimate! Any possibility of a re-review/merge @mistercrunch or @ascott ?

mistercrunch commented:
LGTM. Thanks for bearing with us through many major rebase.

mistercrunch commented:
Sometimes we miss notifications feel free to ping asking for review as often as every other day!
